### PR TITLE
View Discard Pile

### LINKED ISSUE #53: Allow player to look through the discard pile
Related to #52 

Would be nice to look through the discard pile at any point in the game. Perhaps a similar indicator to the deck, with a number showing how many cards. And then you click on it to learn more.

Could reuse the scrying UI for actually flicking through the discard - would just need to turn off the buttons on it and it's good to go.

### PR DESCRIPTION
**Pull request by kilikfan**
First crack at showing the discard pile

### CODE REVIEW

**Comment by kilikfan** [comment]
[Not a fan of what it does to the two buttons, probably an easy fix to make them both buttons](https://paste.pics/c584553a6892138c32e9bda3a2d7380c)
[Hiding the scry buttons seemed to work fine, but I think I need a new button to return to the main play screen as it doesn't currently!](https://paste.pics/d705a1bce0855cd7ec3575bcdf75f669)

First crack at viewing the discard pile, and trying to get some activity going on this project again. Lots of problems with this, but a start at least. As Alex suggested I'm hooking into the scrying UI which seems to work like a dream, just all the intricacies aren't quite there, with a neat exit back to the game, and making the two pile counters look the same. Not a fan of dragging the state into yet another component, will definitely be something I have a big think about sometime soon, to see what we can do to reduce it down a bit, but for now I've done it.

File: lib/model/game_state.dart [hunk 1]
```diff
@@ -50,10 +51,17 @@ class GameState with ChangeNotifier {
     randomiser.shuffle(deck);
   }
 
+  void viewDiscards() {
+    scrying.addAll(discard);
+    showScryButtons = false;
+    notifyListeners();
+  }
```
Comment (ghost) [inline lib/model/game_state.dart]: Thinking it would keep the state cleaner if we didn't reuse the `scrying` variable for this. We could instead introduce a single bool flag, `viewingDiscard`, which we'd use to toggle in the relevant screen. 

In particular, this means we can avoid having `showScryButtons` in game state, which feels slightly polluting because ideally the state shouldn't need to know anything about the UI.

File: lib/screens/play/scrying_page.dart [hunk 2]
```diff
@@ -41,20 +41,22 @@ class ScryingPage extends StatelessWidget {
...
-          ),
-          onPressed: () {
-            state.replaceScryedCard(card, scryOption);
-          },
-          color: scryOption.getColour(),
-          child:
-              Text(scryOption.getDescription(), style: TextStyle(fontSize: 30)),
-        ));
+    return Visibility(
+        visible: state.showScryButtons,
```
Comment (ghost) [inline lib/screens/play/scrying_page.dart]: If we go with my suggested state tweak, we'll need another strategy here. I'd suggest we revert this back to the simple container, and extract the relevant parts to reuse from a new component (`view_discard_page` ?).

So we'd end up with a generic `CardViewer` or some such, which you'd pass a list of Cards and a transformer to translate a card to a widget. The view discard page would pass the discard list, and a simple transformer that just renders the card. The scrying page would pass the scry list, and a transformer that includes the scry buttons.

Then we'll want a switch in the game screen or wherever, to pay attention to the `viewing_discard` bool and show the view discard page appropriately.

It's a bit more code, but it keeps things explicit and separates our responsibilities nicely. Another advantage of separating out a `view_discard_page` is that we can make that one have a 'close' button, which we'd hook into a state method so you can return to the actual game. We wouldn't want this on the scrying page, which would require more boolean magic to toggle on/off with the current approach.

File: lib/screens/play/card/card_pile.dart [hunk 3]
```diff
@@ -1,33 +1,48 @@
...
+                      ? () {
+                          state.viewDiscards();
+                        }
+                      : null,
+                  child: Text(
+                    pileName + ": " + pileSize.toString(),
+                    style: TextStyle(
+                      fontSize: 12.0,
+                      fontWeight: FontWeight.bold,
+                    ),
```
Comment (ghost) [inline lib/screens/play/card/card_pile.dart]: We could refactor this so we can reuse the `Text` component - similar to what we've done for `shape`.

File: lib/screens/play/card/card_pile.dart [hunk 4]
```diff
@@ -1,33 +1,48 @@
...
+                    fontWeight: FontWeight.bold,
+                  ),
+                )
+              : RaisedButton(
+                  shape: shape,
+                  onPressed: pileSize > 0
+                      ? () {
+                          state.viewDiscards();
+                        }
+                      : null,
```
Comment (ghost) [inline lib/screens/play/card/card_pile.dart]: Was weighing up whether it might be better for this logic to live in the state, but one advantage of having it here is that the button greys out in the UI when the pile is empty, which I like. So let's keep it how it is.

File: lib/screens/play/card/card_pile.dart [hunk 5]
```diff
@@ -1,33 +1,48 @@
 import 'package:flutter/material.dart';
+import 'package:treasure_of_the_high_seas/model/game_state.dart';
```
Comment (ghost) [inline lib/screens/play/card/card_pile.dart:3]: Would be good to add some tests for this - there are none currently. Especially now that it has some logic, we can test that e.g:

 - Clicking discard when it's empty does nothing
 - Clicking discard when non-empty has the expected effect on the game state
 - The 'deck' pile doesn't render as a button, and has the correct size rendered

File: lib/model/game_state.dart [hunk 6]
```diff
@@ -50,10 +51,17 @@ class GameState with ChangeNotifier {
     randomiser.shuffle(deck);
   }
 
+  void viewDiscards() {
```
Comment (ghost) [inline lib/model/game_state.dart]: However this ends up looking, we should add tests for it. Feels a bit silly, especially if all it ends up doing is setting a flag, but meh.

Thinking that we might actually end up with this being `toggleDiscardView()`, which would just negate `viewingDiscard`. This way we can hook back into it from an escape button in our discard view (I'm guessing that currently there's no way to get out of this view once you're in it?)

**Comment by ghost** [comment]
> [Not a fan of what it does to the two buttons, probably an easy fix to make them both buttons](https://paste.pics/c584553a6892138c32e9bda3a2d7380c)
> [Hiding the scry buttons seemed to work fine, but I think I need a new button to return to the main play screen as it doesn't currently!](https://paste.pics/d705a1bce0855cd7ec3575bcdf75f669)
> 
> First crack at viewing the discard pile, and trying to get some activity going on this project again. Lots of problems with this, but a start at least. As Alex suggested I'm hooking into the scrying UI which seems to work like a dream, just all the intricacies aren't quite there, with a neat exit back to the game, and making the two pile counters look the same. Not a fan of dragging the state into yet another component, will definitely be something I have a big think about sometime soon, to see what we can do to reduce it down a bit, but for now I've done it.

I don't think it's the end of the world for more components to have visibility of the state - it's the only way they can sensibly drive state changes (for instance, when we add a close button, we'll need to call a method on the state to say we're no longer viewing the discard pile).

IMO, the only painful thing is having to drill it down from the top-level `play_page` all the way into whichever children need it via constructors. That definitely does suck. https://github.com/kilikfan/treasure/issues/38 already exists to try and resolve this, though - we'd move towards a pattern where components are able to "grab" the state as and when they need it (perhaps via a `Provider` ?). So I'm not sure we need the new issue, but I could definitely be missing something!

**Comment by ghost** [comment]
I mean, I guess it's kinda janky that the components have the *whole* state, and could theoretically call any method on there that they wanted. With react/redux, you end up passing functions as props which gets around this quite nicely. So your component would have a `toggleDiscardView` prop, which is just a method. The component then doesn't have a reference to the state, it just has a reference to that single prop - it doesn't know what it does, it just calls it when appropriate. This makes testing quite nice, too, because you can pass a stub function in tests and check it's called (rather than verifying that a potentially complex state operation has taken place). Although from memory mock functions in Dart are a bit painful!

File: lib/screens/play/card/card_pile.dart [hunk 7]
```diff
@@ -1,33 +1,48 @@
...
 
 class CardPile extends StatelessWidget {
```
Comment (JonMasson) [inline lib/screens/play/card/card_pile.dart:5]: Are we likely to have any other card piles than 'deck' and 'discard'?

Comment (kilikfan) [reply lib/screens/play/card/card_pile.dart:5]: Pass!

**Comment by kilikfan** [comment]
Ok, I'm definitely doing something wrong. In my head from scrying I'm passing a List<Widget> from the CardViewer to the ScryingPage and it should grab any extra components and populate the List. That seems to happen but when I come out of there it loses them and components is null, which causes a crash! Also it doesn't run the ViewDiscard stuff at all, so help please!

File: lib/model/game_state.dart [hunk 8]
```diff
@@ -50,6 +51,11 @@ class GameState with ChangeNotifier {
...
+    ViewDiscardPage(this);
```
Comment (ghost) [inline lib/model/game_state.dart]: Ah, yes - so this is at least part of the problem. We're in the state code here, which is separate from the UI - it has no access to the widget tree and no way of rendering components. So, although you've constructed a widget here, it won't actually do anything or render anywhere.

The way the `ScryingPage` gets shown is by using logic in `PlayPage`:

```
if (state.scrying.isNotEmpty) {
    return ScryingPage(state);
  } else if (state.result != null) {
    return GameEndPage(state.result);
  } else { ... }
```

So, all we want to do here is set a new boolean field called `viewingDiscard`. Then update the `PlayPage` to:

```
if (state.scrying.isNotEmpty) {
    return ScryingPage(state);
  } else if (state.viewingDiscard) {
    return ViewDiscard(state);
  } else if (state.result != null) {
    return GameEndPage(state.result);
  } else { ... }
```

This will at least get the `ViewDiscard` widget to show at the appropriate time.

File: lib/screens/play/card/card_pile.dart [hunk 9]
```diff
@@ -1,33 +1,43 @@
...
+        child: Card(
+            shape: shape,
+            child: Padding(
+                padding: EdgeInsets.fromLTRB(4, 4, 4, 4),
+                child: pileName == 'Deck'
+                    ? TextButton(child: text)
+                    : TextButton(
+                        onPressed: pileSize > 0
+                            ? () {
+                                ViewDiscardPage(state);
```
Comment (ghost) [inline lib/screens/play/card/card_pile.dart]: This is a nearer miss than the state example, but still doesn't show anything - because the onPressed is constructing a `ViewDiscardPage` but not actually doing anything with it. Instead, we just want to update the state from here - that will update our boolean, and call `notifyListeners` - which will in turn trigger Flutter to automatically rerender our components.

It's probably easiest to explain with a diagram:

image

File: lib/screens/play/card/card_pile.dart [hunk 10]
```diff
@@ -1,33 +1,42 @@
...
+        child: Card(
+            shape: shape,
+            child: Padding(
+                padding: EdgeInsets.fromLTRB(4, 4, 4, 4),
+                child: pileName == 'Deck'
+                    ? TextButton(child: text, onPressed: null)
+                    : TextButton(
+                        onPressed: pileSize > 0
+                            ? () {
+                                state.toggleDiscardView();
```
Comment (ghost) [inline lib/screens/play/card/card_pile.dart]: We can use the provider stuff here, meaning we don't have to pass the `GameState` to this component at all.

```suggestion
                                context.read<GameState>().toggleDiscardView();
```

File: lib/screens/play/card/card_pile.dart [hunk 11]
```diff
@@ -1,33 +1,42 @@
...
+    return Center(
+        child: Card(
+            shape: shape,
+            child: Padding(
+                padding: EdgeInsets.fromLTRB(4, 4, 4, 4),
+                child: pileName == 'Deck'
+                    ? TextButton(child: text, onPressed: null)
+                    : TextButton(
+                        onPressed: pileSize > 0
+                            ? () {
```
Comment (ghost) [inline lib/screens/play/card/card_pile.dart]: Think we can simplify slightly to a single ternary, just for the `onPressed`:

```
                child: TextButton(
                    onPressed: pileName == 'Discard' && pileSize > 0
                        ? () {
                            context.read<GameState>().toggleDiscardView();
                          }
                        : null,
                    child: text))));
```

File: lib/screens/play/card_viewer.dart [hunk 12]
```diff
@@ -0,0 +1,46 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_swiper/flutter_swiper.dart';
+import 'package:treasure_of_the_high_seas/model/view_mode.dart';
+
+import '../../model/card/card.dart' as Model;
+import 'card/card_display.dart';
+
+class CardViewer extends StatelessWidget {
+  final List<Model.Card> cardList;
+  final ViewMode mode;
```
Comment (ghost) [inline lib/screens/play/card_viewer.dart]: `ViewMode` isn't actually used for anything in here, so we can ditch it :fire:

File: lib/screens/play/card_viewer.dart [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+  Widget build(BuildContext context) {
+    Widget _getCardWidget(Model.Card card) {
+      return Container(
+          padding: EdgeInsets.symmetric(vertical: 5),
+          height: MediaQuery.of(context).size.height * 0.65,
+          child: CardDisplay(
+            card,
+            readOnly: true,
+          ));
+    }
```
Comment (ghost) [inline lib/screens/play/card_viewer.dart]: Would be tempted to pull this outside of `build`, and have it a function sitting directly in `CardViewer`. No functional difference, just a little more readable maybe?

File: lib/screens/play/scrying_page.dart [hunk 14]
```diff
@@ -1,39 +1,27 @@
...
-import 'card/card_display.dart';
+import 'package:treasure_of_the_high_seas/model/view_mode.dart';
 import '../../model/card/card.dart' as Model;
+import 'card_viewer.dart';
 
 class ScryingPage extends StatelessWidget {
+  final GameState state;
+
+  ScryingPage(this.state);
+
```
Comment (ghost) [inline lib/screens/play/scrying_page.dart]: This isn't being used, so we can go back to this not having any properties :smile:

File: lib/screens/play/view_discard_page.dart [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+            side: BorderSide(color: Colors.black, width: 1),
+            borderRadius: BorderRadius.circular(10),
+          ),
+          onPressed: () {
+            // TODO - Provider.of<GameState>(context, listen: false);
+            //Navigator.push(
+              //context,
+              //MaterialPageRoute(builder: (context) => PlayPage("Play Page", state)),
+            //);
+            state.toggleDiscardView(closeWindow: true);
```
Comment (ghost) [inline lib/screens/play/view_discard_page.dart]: Yeah, pretty much what the `TODO` says. I think we can be a bit more concise even, with just:

```
context.read<GameState>().toggleDiscardView(closeWindow: true);
```

And then we don't need to pass `GameState` to this component either :tada:

File: lib/screens/play/view_discard_page.dart [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+  @override
+  Widget build(BuildContext context) {
+    final discard = context.select<GameState, List<Model.Card>>((state) => state.discard);
+
+    return CardViewer(discard,
+        ViewMode.DISCARD, closeButton);
+  }
+
+  List<Widget> closeButton(context, card){
+    return [_getCloseButton(context, card)];
```
Comment (ghost) [inline lib/screens/play/view_discard_page.dart:17]: This current approach totally works, so feel free to do this as a separate change (if at all):

In my head I'd imagined we'd pass an empty list to the `CardViewer` and render our own close button - it'd give us more control over exactly where we want it / how to style it etc. Thinking that a static exit button might make more sense, given that the action isn't tied into the card currently being viewed in any way.

File: lib/model/game_state.dart [hunk 17]
```diff
@@ -50,6 +52,16 @@ class GameState with ChangeNotifier {
     randomiser.shuffle(deck);
   }
 
+  void toggleDiscardView({bool closeWindow = false}) {
+    if (closeWindow) {
+      viewMode = ViewMode.NORMAL;
+    } else {
+      viewMode = ViewMode.DISCARD;
+    }
```
Comment (ghost) [inline lib/model/game_state.dart]: Wondering if, instead of the `ViewMode` enum, we can get away with literally just a boolean here. 

Currently `ViewMode.SCRYING` isn't used (it's passed to `CardPile`, but not used so needn't be), and nor is `ViewMode.NORMAL` except in tests. So perhaps this could become:

```
  void toggleDiscardView() {
    viewingDiscard = !viewingDiscard;
    notifyListeners();
  }
```

File: test/screens/play/card/card_pile_test.dart [hunk 18]
```diff
@@ -0,0 +1,45 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_test/flutter_test.dart';
+import 'package:treasure_of_the_high_seas/model/game_state.dart';
+import 'package:treasure_of_the_high_seas/model/view_mode.dart';
+import 'package:treasure_of_the_high_seas/screens/play/card/card_pile.dart';
+
+import '../../../test_utils.dart';
+
+void main() {
```
Comment (ghost) [inline test/screens/play/card/card_pile_test.dart:8]: These are spot on :grin:

File: test/model/game_state_test.dart [hunk 19]
```diff
@@ -134,6 +134,26 @@ void main() {
...
+      state.nextCard();
+      expect(state.discard, [plunderAWreck]);
+      expect(state.currentCard, aRivalShip);
+
+      state.toggleDiscardView();
+      state.toggleDiscardView(closeWindow: true);
+
+      expect(state.discard, [plunderAWreck]);
+      expect(state.currentCard, aRivalShip);
+    });
```
Comment (ghost) [inline test/model/game_state_test.dart]: Feels a bit of an odd test, because the method we're testing doesn't go near the aspects of the state that we're verifying. Maybe something like this is all we need?:

```
    test('should show the same card after viewing discard pile is finished', () {
      final GameState state = makeGameState();
      expect(state.viewingDiscard, false);

      state.toggleDiscardView();
      expect(state.viewingDiscard, true);

      state.toggleDiscardView();
      expect(state.viewingDiscard, false);
    });
```

**Review by ghost — COMMENTED** [review]
It's all definitely working! Just some tidy-up suggestions and I'm happy to approve. 

I poked around a bit and think the scrying thing you ran into is just an existing bug we can raise a separate issue for. Blame the person who wrote scrying originally! :sweat_smile:

File: lib/screens/play/card/card_pile.dart (hunk 10 above)
Comment (kilikfan) [reply lib/screens/play/card/card_pile.dart]: I get an error "The method 'read' isn't defined for the type 'BuildContext'."

File: lib/screens/play/view_discard_page.dart (hunk 16 above)
Comment (kilikfan) [reply lib/screens/play/view_discard_page.dart:17]: Ok, I'll raise a separate ticket for that, I agree that the close button looks bad swiping with everything.

File: lib/model/game_state.dart (hunk 17 above)
Comment (kilikfan) [reply lib/model/game_state.dart]: Yeah, I did wonder about refactoring scrying to work with the view mode as well. I'd rather that than keep having more and more booleans added, but can change this if you prefer it.

File: lib/screens/play/card/card_pile.dart (hunk 10 above)
Comment (ghost) [reply lib/screens/play/card/card_pile.dart]: You need to import `provider` - it's a bit annoying, for me the import doesn't get suggested so i have to copy and paste it in 😪

File: lib/screens/play/view_discard_page.dart (hunk 16 above)
Comment (kilikfan) [reply lib/screens/play/view_discard_page.dart:17]: Raised as https://github.com/kilikfan/treasure/issues/71

**Comment by kilikfan** [comment]
Crud, changing it to use provider and context.read<GameState> has broken the test in card_pile_test where it clicks the discard pile. Do I just not click it in that test?

**Comment by ghost** [comment]
> Crud, changing it to use provider and context.read has broken the test in card_pile_test where it clicks the discard pile. Do I just not click it in that test?

Nah it's a simple fix, you just need to pass the `game_state` when launching the widget:

```
await tester.launchWidget(child: CardPile(5, 'Discard'), state: state);
```

File: lib/model/game_state.dart (hunk 17 above)
Comment (ghost) [reply lib/model/game_state.dart]: Yeah if we're going to keep `ViewMode` I'd rather we plug it in properly - at the moment it's a bit weird because `SCRYING` is a vestigial thing that's never set or used, which is pretty misleading. Seeing it in the `enum` would definitely make you think it was driving the scrying logic, which it isn't atm.

Maybe we apply the rule of three and hold off for now? Agree that it would be bad to have 5 booleans along the lines of `scrying`, `viewingDiscard`, `viewRemovedFromPlay` etc etc, but there are only 2 right now (and in the scrying case it's not even a boolean, we're able to reuse the `scrying` list - which is quite neat as we need it anyway).

**Review by ghost — APPROVED** [review]
LGTM. Don't mind which direction you want to go in for `ViewMode`, so long as it's all-or-nothing.

**Review by alyssaruth — APPROVED** [review]
